### ISSUE TITLE
Plasma Store Fails when defaulting to pickle

### ISSUE THREAD

**Issue by dmadeka** [ray-project/ray#3881]
I tried to `put` an array of CustomObjects - and got this warning:

```
WARNING: Falling back to serializing objects of type <class 'scipy.interpolate._cubic.CubicSpline'> by using pickle. This may be inefficient.
```

Followed by this error:

```
---------------------------------------------------------------------------
ArrowCapacityError                        Traceback (most recent call last)
<ipython-input-35-4fa2fc614b22> in <module>()
      4 for names, obj in obj_dict.items():
      5     print(names)
----> 6     _pinned_objects[names] = pin_in_object_store(obj)

<ipython-input-34-a2184fde3bd2> in pin_in_object_store(obj)
      1 def pin_in_object_store(obj):
----> 2     obj_id = ray.put(_to_pinnable(obj))
      3     return "{}{}".format(
      4         PINNED_OBJECT_PREFIX, base64.b64encode(obj_id.id()).decode("utf-8"))

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/worker.py in put(value, worker)
   2384         object_id = worker.local_scheduler_client.compute_put_id(
   2385             worker.current_task_id, worker.put_index)
-> 2386         worker.put_object(object_id, value)
   2387         worker.put_index += 1
   2388         return object_id

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/worker.py in put_object(self, object_id, value)
    409         # Serialize and put the object in the object store.
    410         try:
--> 411             self.store_and_register(object_id, value)
    412         except pyarrow.PlasmaObjectExists:
    413             # The object already exists in the object store, so there is no

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/worker.py in store_and_register(self, object_id, value, depth)
    344                     memcopy_threads=self.memcopy_threads,
    345                     serialization_context=self.get_serialization_context(
--> 346                         self.task_driver_id))
    347                 break
    348             except pyarrow.SerializationCallbackError as e:

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/utils.py in _wrapper(*args, **kwargs)
    402                 def _wrapper(*args, **kwargs):
    403                     with self.lock:
--> 404                         return orig_attr(*args, **kwargs)
    405 
    406                 self._wrapper_cache[attr] = _wrapper

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/pyarrow_files/pyarrow/_plasma.pyx in pyarrow._plasma.PlasmaClient.put()

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/pyarrow_files/pyarrow/serialization.pxi in pyarrow.lib.serialize()

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/ray/pyarrow_files/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowCapacityError: BinaryArray cannot contain more than 2147483646 bytes, have 2147486312
```

Is there an easy fix for this?

**Comment by robertnishihara** [comment]
@pcmoritz is currently working on lifting this limitation in Arrow.

One workaround is to define a custom serializer for the `scipy.interpolate._cubic.CubicSpline` type, see https://ray.readthedocs.io/en/latest/api.html#ray.register_custom_serializer. The serializer should take a `CubicSpline` object and perhaps return the relevant parameters as a list of numpy arrays, and the deserializer should do the reverse.

**Comment by robertnishihara** [comment]
If that doesn't work, then you can just pickle the `CubicSpline` object in the custom serializer and then turn the resulting byte string into a numpy array of bytes (because Arrow can serialize numpy arrays of size greater than 2GB (but not strings)).

**Comment by dmadeka** [comment]
That works! I get an even more interesting error:

```
~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/scipy/interpolate/interpolate.py in _evaluate(self, x, nu, extrapolate, out)
    986     def _evaluate(self, x, nu, extrapolate, out):
    987         _ppoly.evaluate(self.c.reshape(self.c.shape[0], self.c.shape[1], -1),
--> 988                         self.x, x, nu, bool(extrapolate), out)
    989 
    990     def derivative(self, nu=1):

_ppoly.pyx in scipy.interpolate._ppoly.evaluate()

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/scipy/interpolate/_ppoly.cpython-36m-x86_64-linux-gnu.so in View.MemoryView.memoryview_cwrapper()

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/scipy/interpolate/_ppoly.cpython-36m-x86_64-linux-gnu.so in View.MemoryView.memoryview.__cinit__()

ValueError: buffer source array is read-only
```
It seems I have to copy the array to avoid this which blows the memory

**Comment by dmadeka** [comment]
btw @robertnishihara When I tried the direct pickle solution I got the error:

```python
def serialize_spline(spline_obj):
    return np.array(pickle.dumps(spline_obj), dtype=bytes)

def deserialize_spline(spline_obj):
    return pickle.loads(spline_obj)

obj_id = ray.put(_to_pinnable(spline_array))

ValueError: Changing the dtype of a 0d array is only supported if the itemsize is unchanged
```

**Comment by robertnishihara** [comment]
Maybe try

```python
def serialize_spline(spline_obj):
    return np.array(bytearray(pickle.dumps(spline_obj)), dtype=np.uint8)

def deserialize_spline(serialized_obj):
    return pickle.loads(serialized_obj)
```

**Comment by robertnishihara** [comment]
@dmadeka, yes `pickle.loads` will create a copy. If you want to do it without copying the array, then you'll need to define a custom serializer that takes the spline object and returns primitive Python types (e.g., a list containing the array and other relevant parameters). You'll also need a custom deserializer that takes the primitive Python types and reconstructs the spline object. Does that make sense?

**Comment by dmadeka** [comment]
Yeah, @robertnishihara that makes sense. It was blocked by this - https://github.com/scipy/scipy/issues/9740

Now that the PR is merged, ill update this thread using a nightly build of scipy

**Comment by edoakes** [comment]
Stale - please open new issue if still relevant
